### PR TITLE
Fix a few forward declarations under utilities folder

### PR DESCRIPTION
**Pull request by cngzhnp**
This changeset contains a few fix about forward declaration of classes. There a lot of fixes needed for utilities class but I'd prefer to fix them not at once.

Please review these changes.

### CODE REVIEW

**Comment by riversand963** [comment]
Thanks @cngzhnp for the PR.
Could you include more information in the PR description to explain the fixes?
IIRC, is this related to https://github.com/facebook/rocksdb/pull/4393?

**Comment by cngzhnp** [comment]
Hello @riversand963,

Yes, you're right. My main goal is, solve such unnecessary inclusions and improve build time of all system in meanwhile. You might think of this is the first part of my changes as I told in #4393. 

I want to improve it folder by folder (maybe subfolder). IMHO, it would be clear to do core review with small changes, if it is OK for you as well. What is your feeling?

**Comment by siying** [comment]
@riversand963 I think it's generally good to avoid forward declaration. Google C++ Style, which we are following in this code base, actually discourages it: https://google.github.io/styleguide/cppguide.html#Forward_Declarations

**Comment by cngzhnp** [comment]
I disagree with @siying. To avoid forward declarations only make sense in this : "Try to avoid forward declarations of entities defined in another project"

Otherwise, forward declaration cannot be done if there is an implementation in header file so there is no recompilation problems when you use it. As far as I see, there is no template classes or single-included header files in this project.

Therefore, in standard library there is no forward declaration except (iosfwd.h) so it cannot be done and not recommended. 

To sum up, I could not see any side effect not to use forward declarations if those are defined by your project developers.

**Comment by cngzhnp** [comment]
If there will be no action to take on this PR, I will close it.
